### PR TITLE
only directly query `.nuspec` files from nuget and azure devops

### LINKED ISSUE #8837: Nuget doesn't work with AWS Codeartifact private repo
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package ecosystem

nuget

### Package manager version

3

### Language version

_No response_

### Manifest location and content before the Dependabot update

n/a

### dependabot.yml content

```
version: 2
registries:
  codeartifact:
    type: nuget-feed
    url: https://landr-**redacted**.d.codeartifact.us-east-1.amazonaws.com/nuget/landr/v3/index.json
    token: ${{secrets.CODEARTIFACT_TOKEN}}
updates:
  - package-ecosystem: "nuget"
    directory: "/"
    registries:
      - codeartifact
    allow:
      - dependency-name: "LANDR.MicroServices.Core*"
    schedule:
      interval: "weekly"
```

### Updated dependency

_No response_

### What you expected to see, versus what you actually saw

When running against an AWS Codeartifact nuget repository, dependabot fails to fetch `.nuspec` files, and ends up with 'unknown_error'.

``` 
  proxy | 2024/01/17 22:16:33 [034] GET https://landr-*redacted*.d.codeartifact.us-east-1.amazonaws.com:443/nuget/landr/v3/registration4/landr.microservices.core.servicebootstrapper/index.json
  proxy | 2024/01/17 22:16:33 [034] * authenticating nuget feed request (host: landr-*redacted*.d.codeartifact.us-east-1.amazonaws.com, bearer auth)
  proxy | 2024/01/17 22:16:33 [034] 200 https://landr-*redacted*.d.codeartifact.us-east-1.amazonaws.com:443/nuget/landr/v3/registration4/landr.microservices.core.servicebootstrapper/index.json
[...]
  proxy | 2024/01/17 22:16:38 [078] GET https://landr-*redacted*.d.codeartifact.us-east-1.amazonaws.com:443/nuget/landr/v3-flatcontainer/landr.microservices.core.servicebootstrapper/6.15.1/landr.microservices.core.servicebootstrapper.nuspec
  proxy | 2024/01/17 22:16:38 [078] 404 https://landr-*redacted*.d.codeartifact.us-east-1.amazonaws.com:443/nuget/landr/v3-flatcontainer/landr.microservices.core.servicebootstrapper/6.15.1/landr.microservices.core.servicebootstrapper.nuspec
[...]
updater | 2024/01/17 22:16:45 ERROR <job_775557849> Error processing LANDR.MicroServices.Core.Redis (NoMethodError)
updater | 2024/01/17 22:16:45 ERROR <job_775557849> undefined method `fetch' for nil:NilClass
updater | 
updater |           latest_version: preferred_resolvable_version_details.fetch(:version)&.to_s,
updater |                                                               ^^^^^^
updater | 2024/01/17 22:16:45 ERROR <job_775557849> /home/dependabot/nuget/lib/dependabot/nuget/update_checker.rb:47:in `updated_requirements'
updater | 2024/01/17 22:16:45 ERROR <job_775557849> /home/dependabot/nuget/lib/dependabot/nuget/update_checker.rb:101:in `updated_dependencies_after_full_unlock'
updater | 2024/01/17 22:16:45 ERROR <job_775557849> /home/dependabot/common/lib/dependabot/update_checkers/base.rb:61:in `updated_dependencies'
updater | 2024/01/17 22:16:45 ERROR <job_775557849> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/update_all_versions.rb:99:in `check_and_create_pull_request'
updater | 2024/01/17 22:16:45 ERROR <job_775557849> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/update_all_versions.rb:60:in `check_and_create_pr_with_error_handling'
updater | 2024/01/17 22:16:45 ERROR <job_775557849> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/update_all_versions.rb:35:in `block in perform'
updater | 2024/01/17 22:16:45 ERROR <job_775557849> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/update_all_versions.rb:35:in `each'
updater | 2024/01/17 22:16:45 ERROR <job_775557849> /home/dependabot/dependabot-updater/lib/dependabot/updater/operations/update_all_versions.rb:35:in `perform'
updater | 2024/01/17 22:16:45 ERROR <job_775557849> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:45:in `run'
updater | 2024/01/17 22:16:45 ERROR <job_775557849> /home/dependabot/dependabot-updater/lib/dependabot/update_files_command.rb:43:in `perform_job'
updater | 2024/01/17 22:16:45 ERROR <job_775557849> /home/dependabot/dependabot-updater/lib/dependabot/base_command.rb:37:in `run'
updater | 2024/01/17 22:16:45 ERROR <job_775557849> bin/update_files.rb:24:in `<main>'
updater | 2024/01/17 22:16:45 INFO <job_775557849> Finished job processing
updater | 2024/01/17 22:16:45 INFO Results:
updater | Dependabot encountered '6' error(s) during execution, please check the logs for more details.
updater | +--------------------------------------------------------------+
updater | |                Dependencies failed to update                 |
updater | +----------------------------------------------+---------------+
updater | | LANDR.MicroServices.Core.ServiceBootstrapper | unknown_error |
updater | | LANDR.MicroServices.Core.Lambdas             | unknown_error |
updater | | LANDR.MicroServices.Core.EntityFrameworkCore | unknown_error |
updater | | LANDR.MicroServices.Core.TestHelpers         | unknown_error |
updater | | LANDR.MicroServices.Core.DataMigrator        | unknown_error |
updater | | LANDR.MicroServices.Core.Redis               | unknown_error |
updater | +----------------------------------------------+---------------+
```
(shown here only with `LANDR.MicroServices.Core.ServiceBootstrapper`, but all other packages exhibit the same 404 issue)

I believe this stems from https://github.com/dependabot/dependabot-core/blob/715b2548fd1792c85809a72bd1df7ab05a423b2b/nuget/lib/dependabot/nuget/update_checker/nuspec_fetcher.rb#L39, where the base URL is generated rather than being obtained from the `index.json` that we specify in dependabot.yml.

This index.json contains the following :
```
    {
      "@id": "https://landr-*redacted*.d.codeartifact.us-east-1.amazonaws.com/nuget/landr/v3/flatcontainer",
      "@type": "PackageBaseAddress/3.0.0"
    }
```
(note the `/flatcontainer`, vs `-flatcontainer`)

This might be related to other nuget issues, such as https://github.com/dependabot/dependabot-core/issues/8759

### Native package manager behavior

_No response_

### Images of the diff or a link to the PR, issue, or logs

_No response_

### Smallest manifest that reproduces the issue

_No response_

### PR DESCRIPTION
**Pull request by brettfo**
Previously the NuGet updater would directly query every package feed _except_ Azure DevOps for the `.nuspec` contents, but this didn't account for both GitHub and AWS package feeds that return 404 on a `.nuspec`.

The fix is to _only_ directly pull the `.nuspec` file from the feeds that support it, specifically NuGet.org and Azure DevOps; all other feeds we pull the `.nupkg` then manually extract the `.nuspec`.  If/when we find more package feeds that _do_ allow direct pulling of the `.nuspec` files, we can add them to the allow list.

This will fix issues like the following, where the `.nuspec` couldn't be queried which meant we weren't able to parse any meaningful version numbers:

```
  proxy | 2024/01/17 22:16:38 [078] GET https://nuget.example.com/some.package/1.2.3/some.package.nuspec
  proxy | 2024/01/17 22:16:38 [078] 404 https://nuget.example.com/some.package/1.2.3/some.package.nuspec
...
updater | 2024/01/17 22:16:45 ERROR <job_775557849> undefined method `fetch' for nil:NilClass
updater | 
updater |           latest_version: preferred_resolvable_version_details.fetch(:version)&.to_s,
```

This also fixes an issue with GitHub package feeds which don't include the `listed` property.

Fixes #8837 (and likely many other issues; I'll have to review each of them)

### CODE REVIEW

**Review by abdulapopoola — APPROVED** [review]

File: nuget/lib/dependabot/nuget/update_checker/nuspec_fetcher.rb [hunk 1]
```diff
@@ -26,17 +26,9 @@ def self.fetch_nuspec_from_repository(repository_details, package_id, package_ve
...
-                                                                         package_version)
-          return if package_data.nil?
-
-          nuspec_string = extract_nuspec(package_data, package_id)
-          nuspec_xml = Nokogiri::XML(nuspec_string)
-        else
+        if feed_supports_nuspec_download?(feed_url)
           # we can use the normal nuget apis to get the nuspec and list out the dependencies
-          base_url = feed_url.gsub("/index.json", "-flatcontainer")
+          base_url = repository_details[:base_url].delete_suffix("/")
```
Comment (brettfo) [inline nuget/lib/dependabot/nuget/update_checker/nuspec_fetcher.rb:31]: This change is what allows us to pull `.nuspec` files directly from Azure DevOps.

File: nuget/lib/dependabot/nuget/nuget_client.rb [hunk 2]
```diff
@@ -68,7 +68,11 @@ def self.get_package_versions(dependency_name, repository_details)
             # inlined entries
             items.each do |item|
               catalog_entry = item["catalogEntry"]
-              if catalog_entry["listed"] == true
+
+              # a package is considered listed if the `listed` property is either `true` or missing
+              listed_property = catalog_entry["listed"]
```
Comment (ryanbrandenburg) [inline nuget/lib/dependabot/nuget/nuget_client.rb:73]: Ah, good catch!

File: nuget/spec/dependabot/nuget/nuget_client_spec.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
+# typed: false
+# frozen_string_literal: true
+
+require "spec_helper"
+require "dependabot/nuget/nuget_client"
+
+RSpec.describe Dependabot::Nuget::NugetClient do
```
Comment (ryanbrandenburg) [inline nuget/spec/dependabot/nuget/nuget_client_spec.rb:7]: Good to start this test class, I was relying on the using areas to test the functionality, but as it starts touching more spots that stops being as good of an idea.

**Review by ryanbrandenburg — APPROVED** [review]

File: nuget/lib/dependabot/nuget/nuget_client.rb (hunk 2 above)
Comment (brettfo) [reply nuget/lib/dependabot/nuget/nuget_client.rb:73]: This part may or may not have taken me 2 hours to figure out.

**Review by edgarrs — APPROVED** [review]

File: nuget/lib/dependabot/nuget/update_checker/nuspec_fetcher.rb [hunk 4]
```diff
@@ -49,20 +41,30 @@ def self.fetch_nuspec_from_repository(repository_details, package_id, package_ve
...
+        feed_rexexs = [
+          # nuget
+          %r{https://api\.nuget\.org/v3/index\.json},
+          # azure devops
           %r{https://pkgs\.dev\.azure\.com/(?<organization>[^/]+)/(?<project>[^/]+)/_packaging/(?<feedId>[^/]+)/nuget/v3/index\.json},
           %r{https://pkgs\.dev\.azure\.com/(?<organization>[^/]+)/_packaging/(?<feedId>[^/]+)/nuget/v3/index\.json(?<project>)},
           %r{https://(?<organization>[^\.\/]+)\.pkgs\.visualstudio\.com/_packaging/(?<feedId>[^/]+)/nuget/v3/index\.json(?<project>)}
         ]
-        azure_devops_regexs.any? { |reg| reg.match(feed_url) }
+        feed_rexexs.any? { |reg| reg.match(feed_url) }
```
Comment (JoeRobich) [inline nuget/lib/dependabot/nuget/update_checker/nuspec_fetcher.rb:67]: Is `feed_rexexs` a typo of does it mean something in ruby? =)

**Review by JoeRobich — APPROVED** [review]

File: nuget/lib/dependabot/nuget/update_checker/nuspec_fetcher.rb (hunk 4 above)
Comment (brettfo) [reply nuget/lib/dependabot/nuget/update_checker/nuspec_fetcher.rb:67]: Bold of you to assume I can spell :).  I'll fix the typo, good catch.

**Pull request merged**